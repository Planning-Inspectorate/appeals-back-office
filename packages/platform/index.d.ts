import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
