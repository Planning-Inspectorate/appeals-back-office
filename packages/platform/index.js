export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
