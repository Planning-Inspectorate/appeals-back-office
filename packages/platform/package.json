{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
