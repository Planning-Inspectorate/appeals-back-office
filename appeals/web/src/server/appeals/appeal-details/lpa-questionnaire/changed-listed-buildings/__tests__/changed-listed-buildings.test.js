import { parseHtml } from '@pins/platform';
import supertest from 'supertest';
import { appealData, lpaQuestionnaireData } from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';
import nock from 'nock';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';
const appealId = appealData.appealId;
const lpaQuestionnaireId = appealData.lpaQuestionnaireId;
const validListedBuilding = '1234567';
const invalidListedBuilding = '9999999';

describe('changed-listed-buildings', () => {
	beforeEach(() => {
		installMockApi();
		nock('http://test/')
			.get(`/appeals/${appealId}/lpa-questionnaires/${lpaQuestionnaireId}`)
			.reply(200, {
				...lpaQuestionnaireData
			});
		nock('http://test/').get(`/appeals/listed-buildings/${validListedBuilding}`).reply(200, {});
		nock('http://test/').get(`/appeals/listed-buildings/${invalidListedBuilding}`).reply(404, {});
	});
	afterEach(teardown);

	describe('GET /add', () => {
		it('should render the add changed listed building page', async () => {
			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add`
			);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- add changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');
		});
	});

	describe('POST /add', () => {
		it('should re-render the add changed listed building page with an error when listed building is empty', async () => {
			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add`
				)
				.send({});
			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- add changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Provide a listed building entry list number</a>');
		});

		it('should re-render the add changed listed building page with an error when listed building is less than seven digits', async () => {
			const invalidData = {
				changedListedBuilding: '12345'
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add`
				)
				.send(invalidData);
			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- add changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Listed building entry number must be 7 digits</a>');
		});

		it('should re-render the add changed listed building page with an error when listed building is more than seven digits', async () => {
			const invalidData = {
				changedListedBuilding: '12345678'
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add`
				)
				.send(invalidData);
			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- add changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Listed building entry number must be 7 digits</a>');
		});

		it('should re-render the add changed listed building page with an error when listed building contains a special character', async () => {
			const invalidData = {
				changedListedBuilding: '*123456'
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add`
				)
				.send(invalidData);
			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- add changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Listed building entry number must be 7 digits</a>');
		});

		it('should re-render the add changed listed building page with an error when listed building contains a letter', async () => {
			const invalidData = {
				changedListedBuilding: 'A123456'
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add`
				)
				.send(invalidData);

			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- add changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Listed building entry number must be 7 digits</a>');
		});

		it('should re-render the add changed listed building page with an error when listed building is seven digits but not in historic england', async () => {
			const invalidData = {
				changedListedBuilding: invalidListedBuilding
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add`
				)
				.send(invalidData);
			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- add changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Enter a real listed building entry number</a>');
		});

		it('should redirect to the check-and-confirm page when data is valid', async () => {
			const validData = {
				changedListedBuilding: validListedBuilding
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add`
				)
				.send(validData);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/1/lpa-questionnaire/1/changed-listed-buildings/add/check-and-confirm'
			);
		});
	});

	describe('GET /add/check-and-confirm', () => {
		it('should render the add changed listed building check and confirm page', async () => {
			const validData = {
				changedListedBuilding: validListedBuilding
			};

			await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add`
				)
				.send(validData);

			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add/check-and-confirm`
			);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('Check details and add changed listed building</h1>');

			const summaryListHtml = parseHtml(response.text, {
				rootElement: '.govuk-summary-list',
				skipPrettyPrint: true
			}).innerHTML;
			expect(summaryListHtml).toContain('Changed listed building</dt>');
			expect(summaryListHtml).toContain(
				`<a href="https://historicengland.org.uk/listing/the-list/list-entry/${validListedBuilding}" class="govuk-link" target="_blank">${validListedBuilding}</a>`
			);
			expect(elementInnerHtml).toContain('Add changed listed building</button>');
		});
	});

	describe('POST /add/check-and-confirm', () => {
		it('should redirect to the lpa questionnaire page', async () => {
			const validData = {
				changedListedBuilding: validListedBuilding
			};
			nock('http://test/').post(`/appeals/${appealId}/listed-buildings`).reply(200, {});

			await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add`
				)
				.send(validData);

			const response = await request.post(
				`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/add/check-and-confirm`
			);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				`Found. Redirecting to /appeals-service/appeal-details/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}`
			);
		});
	});

	describe('GET /manage', () => {
		it('should render the manage changed listed buildings page', async () => {
			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/manage`
			);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toContain('Changed listed buildings</h1>');

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;
			expect(unprettifiedElement).toContain('Listed building number</th>');
			expect(unprettifiedElement).toContain('Action</th>');
			expect(unprettifiedElement).toContain(
				'Change<span class="govuk-visually-hidden"> listed building 1234569</span></a>'
			);
			expect(unprettifiedElement).toContain(
				'Remove<span class="govuk-visually-hidden"> listed building 1234569</span></a>'
			);
		});
	});

	describe('GET /change/:listedBuildingId', () => {
		it('should render the change listed building page', async () => {
			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1`
			);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- update changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');
		});
	});

	describe('POST /change/:listedBuildingId', () => {
		it('should re-render the change listed building page when the listed building is empty', async () => {
			const invalidData = {
				changedListedBuilding: null
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1`
				)
				.send(invalidData);

			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;
			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- update changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');
		});

		it('should re-render the change changed listed building page with an error when listed building is less than seven digits', async () => {
			const invalidData = {
				changedListedBuilding: '12345'
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1`
				)
				.send(invalidData);
			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- update changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Listed building entry number must be 7 digits</a>');
		});

		it('should re-render the change changed listed building page with an error when listed building is more than seven digits', async () => {
			const invalidData = {
				changedListedBuilding: '12345678'
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1`
				)
				.send(invalidData);
			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- update changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Listed building entry number must be 7 digits</a>');
		});

		it('should re-render the change changed listed building page with an error when listed building contains a special character', async () => {
			const invalidData = {
				changedListedBuilding: '*123456'
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1`
				)
				.send(invalidData);
			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- update changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Listed building entry number must be 7 digits</a>');
		});

		it('should re-render the change changed listed building page with an error when listed building contains a letter', async () => {
			const invalidData = {
				changedListedBuilding: 'A123456'
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1`
				)
				.send(invalidData);

			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- update changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Listed building entry number must be 7 digits</a>');
		});

		it('should re-render the change changed listed building page with an error when listed building is seven digits but not in historic england', async () => {
			const invalidData = {
				changedListedBuilding: invalidListedBuilding
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1`
				)
				.send(invalidData);
			expect(response.statusCode).toBe(200);
			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('- update changed listed building</span>');
			expect(elementInnerHtml).toContain('Changed listed building entry number</h1>');
			expect(elementInnerHtml).toContain('This is a 7 digit number from Historic England</div>');

			const errorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(errorSummaryHtml).toContain('There is a problem</h2>');
			expect(errorSummaryHtml).toContain('Enter a real listed building entry number</a>');
		});

		it('should redirect to to the check and confirm page if the data is valid', async () => {
			const validData = {
				changedListedBuilding: validListedBuilding
			};

			const response = await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1`
				)
				.send(validData);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/1/lpa-questionnaire/1/changed-listed-buildings/change/1/check-and-confirm'
			);
		});
	});

	describe('GET /change/:listedBuildingId/check-and-confirm', () => {
		it('should render the check and confirm page', async () => {
			const validData = {
				changedListedBuilding: validListedBuilding
			};
			await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1`
				)
				.send(validData);
			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1/check-and-confirm`
			);

			expect(response.statusCode).toBe(200);

			const elementInnerHtml = parseHtml(response.text).innerHTML;
			expect(elementInnerHtml).toMatchSnapshot();

			expect(elementInnerHtml).toContain('Check details and update changed listed building</h1>');
			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true }).innerHTML;
			expect(unprettifiedElement).toContain('Changed listed building</dt>');
			expect(unprettifiedElement).toContain(
				`<a href="https://historicengland.org.uk/listing/the-list/list-entry/${validListedBuilding}" class="govuk-link" target="_blank">${validListedBuilding}</a>`
			);
			expect(elementInnerHtml).toContain('Update changed listed building</button>');
		});
	});

	describe('POST /change/:listedBuildingId/check-and-confirm', () => {
		it('should redirect to lpa questionnaire', async () => {
			const validData = {
				changedListedBuilding: validListedBuilding
			};

			nock('http://test/').patch(`/appeals/${appealId}/listed-buildings/1`).reply(200, {});

			await request
				.post(
					`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1`
				)
				.send(validData);
			const response = await request.post(
				`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/change/1/check-and-confirm`
			);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/1/lpa-questionnaire/1'
			);
		});
	});

	describe('GET /remove/:listedBuildingId', () => {
		it('should render the remove changed listed building page', async () => {
			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/remove/1`
			);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(response.statusCode).toBe(200);
			expect(elementInnerHtml).toContain(
				'Confirm that you want to remove the changed listed building</h1>'
			);
			expect(elementInnerHtml).toContain('Remove changed listed building</button>');
		});
	});

	describe('POST /remove/:listedBuildingId', () => {
		it('should redirect to manage listed building', async () => {
			nock('http://test/').delete(`/appeals/${appealId}/listed-buildings`).reply(200, {});
			const response = await request.post(
				`${baseUrl}/${appealId}/lpa-questionnaire/${lpaQuestionnaireId}/changed-listed-buildings/remove/1`
			);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				`Found. Redirecting to /appeals-service/appeal-details/1/lpa-questionnaire/1`
			);
		});
	});
});
