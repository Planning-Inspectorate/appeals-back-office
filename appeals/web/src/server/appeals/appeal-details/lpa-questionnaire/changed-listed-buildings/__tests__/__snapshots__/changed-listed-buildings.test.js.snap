// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changed-listed-buildings GET /add should render the add changed listed building page 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062 - add changed listed building</span>
            <h1             class="govuk-heading-l">Changed listed building entry number</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group">
                            <div id="changedListedBuilding-hint" class="govuk-hint">This is a 7 digit number from Historic England</div>
                            <input class="govuk-input"
                            id="changedListedBuilding" name="changedListedBuilding" type="text" value=""
                            aria-describedby="changedListedBuilding-hint">
                        </div>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Continue</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;

exports[`changed-listed-buildings GET /add/check-and-confirm should render the add changed listed building check and confirm page 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">Check details and add changed listed building</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <dl class="govuk-summary-list govuk-!-margin-bottom-9">
                <div class="govuk-summary-list__row"><dt class="govuk-summary-list__key"> Changed listed building</dt>
                    <dd class="govuk-summary-list__value"><a href="https://historicengland.org.uk/listing/the-list/list-entry/12345"
                        class="govuk-link" target="_blank">12345</a>
                    </dd>
                    <dd class="govuk-summary-list__actions"><a class="govuk-link" href="/appeals-service/appeal-details/1/lpa-questionnaire/1/changed-listed-buildings/add">Change<span class="govuk-visually-hidden"> listed building number</span></a>
                    </dd>
                </div>
            </dl>
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="post" novalidate="novalidate">
                        <button type="submit" class="govuk-button" data-module="govuk-button">Add changed listed building</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;

exports[`changed-listed-buildings GET /change/:listedBuildingId should render the change listed building page 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062 - update changed listed building</span>
            <h1             class="govuk-heading-l">Changed listed building entry number</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group">
                            <div id="changedListedBuilding-hint" class="govuk-hint">This is a 7 digit number from Historic England</div>
                            <input class="govuk-input"
                            id="changedListedBuilding" name="changedListedBuilding" type="text" value="123456"
                            aria-describedby="changedListedBuilding-hint">
                        </div>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Continue</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;

exports[`changed-listed-buildings GET /change/:listedBuildingId/check-and-confirm should render the check and confirm page 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">Check details and update changed listed building</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <dl class="govuk-summary-list govuk-!-margin-bottom-9">
                <div class="govuk-summary-list__row"><dt class="govuk-summary-list__key"> Changed listed building</dt>
                    <dd class="govuk-summary-list__value"><a href="https://historicengland.org.uk/listing/the-list/list-entry/12345"
                        class="govuk-link" target="_blank">12345</a>
                    </dd>
                    <dd class="govuk-summary-list__actions"><a class="govuk-link" href="/appeals-service/appeal-details/1/lpa-questionnaire/1/changed-listed-buildings/change/1">Change<span class="govuk-visually-hidden"> changed listed building</span></a>
                    </dd>
                </div>
            </dl>
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="post" novalidate="novalidate">
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Update changed listed building</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;

exports[`changed-listed-buildings GET /remove/:listedBuildingId should render the remove changed listed building page 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">Confirm that you want to remove the changed listed building</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <p class="govuk-body"><a href="https://historicengland.org.uk/listing/the-list/list-entry/123456"
                            class="govuk-link" target="_blank">123456</a>
                        </p>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Remove changed listed building</button>
                    </form>
                </div>
            </div>
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <p class="govuk-body"><a href="/appeals-service/appeal-details/1/lpa-questionnaire/1/changed-listed-buildings/manage"
                        class="govuk-link" target="_blank">Cancel</a>
                    </p>
                </div>
            </div>
        </div>
    </div>
</main>"
`;

exports[`changed-listed-buildings POST /add should re-render the add changed listed building page with an error when listed building is empty 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-error-summary" data-module="govuk-error-summary">
                <div role="alert">
                    <h2 class="govuk-error-summary__title"> There is a problem</h2>
                    <div class="govuk-error-summary__body">
                        <ul class="govuk-list govuk-error-summary__list">
                            <li><a href="#changed-listed-building">Provide a listed building entry list number</a>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062 - add changed listed building</span>
            <h1             class="govuk-heading-l">Changed listed building entry number</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group">
                            <div id="changedListedBuilding-hint" class="govuk-hint">This is a 7 digit number from Historic England</div>
                            <input class="govuk-input"
                            id="changedListedBuilding" name="changedListedBuilding" type="text" value=""
                            aria-describedby="changedListedBuilding-hint">
                        </div>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Continue</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;

exports[`changed-listed-buildings POST /change/:listedBuildingId should re-render the change listed building page when the data is invalid 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-error-summary" data-module="govuk-error-summary">
                <div role="alert">
                    <h2 class="govuk-error-summary__title"> There is a problem</h2>
                    <div class="govuk-error-summary__body">
                        <ul class="govuk-list govuk-error-summary__list">
                            <li><a href="#changed-listed-building">Provide a listed building entry list number</a>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062 - update changed listed building</span>
            <h1             class="govuk-heading-l">Changed listed building entry number</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group">
                            <div id="changedListedBuilding-hint" class="govuk-hint">This is a 7 digit number from Historic England</div>
                            <input class="govuk-input"
                            id="changedListedBuilding" name="changedListedBuilding" type="text" value="123456"
                            aria-describedby="changedListedBuilding-hint">
                        </div>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Continue</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;
